patches/0007-Add-informative-backend-configuration-errors.patch

```diff
@@ -0,0 +1,322 @@
++		for _, c := range comment.List {
++			if strings.HasPrefix(c.Text, "//go:build ") {
++				if c, err := constraint.Parse(c.Text); err == nil {
++					return c, nil
++				}
++			}
++		}
++	}
++	return nil, fmt.Errorf("no build constraint found in: %v", filename)
++}
+diff --git a/src/runtime/backenderr_unknown.go b/src/runtime/backenderr_unknown.go
+new file mode 100644
+index 00000000000000..7ea25d50924fde
+--- /dev/null
++++ b/src/runtime/backenderr_unknown.go
+@@ -0,0 +1,17 @@
++// Copyright 2023 The Go Authors. All rights reserved.
++// Use of this source code is governed by a BSD-style
++// license that can be found in the LICENSE file.
++
++// This file is generated by backenderrs.go using the mkbackenderrs.go tool. DO NOT EDIT.
++
++//go:build !((!boringcrypto && !goexperiment.opensslcrypto && !goexperiment.cngcrypto) || (!linux && !windows) || (linux && !cgo) || android || cmd_go_bootstrap || msan) && !(goexperiment.boringcrypto && linux && cgo && amd64 && !android && !cmd_go_bootstrap && !msan) && !(goexperiment.cngcrypto && windows && !cmd_go_bootstrap && !msan) && !(goexperiment.opensslcrypto && linux && cgo && !android && !cmd_go_bootstrap && !msan)
++
++package runtime
++
++func init() {
++	`
++	The current combination of build tags removes the use of built-in Go cryptography but doesn't enable a crypto backend.
++	Make sure the crypto backend selected by GOEXPERIMENT or '-tags' is supported on the target OS (GOOS).
++	For more information, visit https://github.com/microsoft/go/tree/microsoft/main/eng/doc/fips
++	`
++}
+diff --git a/src/runtime/backenderrs.go b/src/runtime/backenderrs.go
+new file mode 100644
+index 00000000000000..30fe5c2b0166bf
+--- /dev/null
++++ b/src/runtime/backenderrs.go
+@@ -0,0 +1,24 @@
++// Copyright 2023 The Go Authors. All rights reserved.
++// Use of this source code is governed by a BSD-style
++// license that can be found in the LICENSE file.
++
++package runtime
++
++// Generate Go files that detect issues with the backend selection and report an
++// error at compile time. This is done in the runtime package rather than the
++// crypto/internal/backend package to make sure these helpful errors will show
++// up. If the files were in the backend package, DuplicateDecl and other errors
++// would show up first, causing these informative errors to be skipped because
++// there are too many total errors already reported.
++//
++// The errors would also show up if we put these files in the crypto package
++// rather than the runtime package. Crypto is imported before the backend
++// backage, so the errors would show up. However, then these errors would show
++// up only if the Go program is using crypto. This could cause a confusing
++// situation: if the user has a misconfigured backend and doesn't use crypto in
++// their Go app, they will not get any errors. If they start using crypto later,
++// they would only then get an error, but the cause would be much less apparent.
```

Comment:
This is what the tests are (basically) running up against. The tests were running fine before this change because they don't actually perform crypto. This meant `crypto/internal/backend` wasn't being imported. But with `backenderr_unknown.go` being plopped in `runtime`, it's always imported and causes a few tests that perform unusual builds to fail.

### Boring Linux job failure

`cmd/internal/testdir Test/codegen/copy.go` ends up running something like:

```
GOEXPERIMENT=boringcrypto GOOS=linux GOARCH=386 CGO_ENABLED=1 ../go/bin/go build .
```

The boring backend has the `amd64` constraint, and `nobackend.go` doesn't have a condition for `!(boringcrypto && !amd64)`, leaving this case unhandled, which triggers `backenderr_unknown.go` and a build/test failure.

### CNG Windows job failure

`cmd/internal/obj/x86 TestVexEvexPCrelative` tries to run something like this:

```
GOEXPERIMENT=cngcrypto GOOS=linux GOARCH=amd64 ../go/bin/go build .
```

CNG doesn't work on Linux, of course, but `nobackend.go` doesn't have a condition that lets it fill in.

### OpenSSL Linux job (pass)

There wasn't a problem with opensslcrypto because its build constraints are more permissive (allows `386`). Also, if a test needs to do something and doesn't care about the platform (like `TestVexEvexPCrelative`), `linux/amd64` is typically the default choice.

---

It was intentional to make the build fail more proactively than it used to (described in this comment), but I think I need to remove this particular error detector to let the tests succeed.

This did make me realize that maybe I could take the code that generates the `backenderr_unknown.go` constraint and instead have it test/write the constraint for `nobackend.go`. I'll file an issue to think over whether that's a good idea. 😄

Reply:
The end result after fixing CI:

* `GOOS=linux GOEXPERIMENT=systemcrypto,boringcrypto go build .` 
  "The boring and openssl backends are both enabled, but they are mutually exclusive. ..."
* `GOOS=freebsd GOEXPERIMENT=systemcrypto go build .`
  "The systemcrypto feature is enabled, but it was unable to enable an appropriate crypto backend for the target GOOS."
* `GOOS=linux GOEXPERIMENT=cngcrypto go build .` 
  ```
  # crypto/internal/backend
  ../go/src/crypto/internal/backend/common.go:41:5: undefined: Enabled
  ../go/src/crypto/internal/backend/common.go:64:5: undefined: Enabled
  ```

Opened https://github.com/microsoft/go/issues/953 to go into the last case further.